### DISCUSSION TITLE
Kalman Filter Debugging, and Lighthouse Configuration Change

### DISCUSSION THREAD

**Discussion by willgraham345** [bitcraze/discussions#861] [Q&A]
Hi all! I've been working (somewhat intermittently) on a project involving mobile lighthouse basestations with the [lighthouse deck](https://www.bitcraze.io/products/lighthouse-positioning-deck/) mounted in a different configuration. I've listed an issue or two regarding the topic [here](https://github.com/bitcraze/crazyflie-firmware/issues/1197). Arnaud Taffanel was nice enough to point me to the [sensor position order](https://github.com/bitcraze/crazyflie-firmware/blob/9e7e697906eced761d5fd27be4f5d866654da250/src/modules/src/lighthouse/lighthouse_position_est.c#L207-L214), as well as the [yaw estimation function](https://github.com/bitcraze/crazyflie-firmware/blob/master/src/utils/src/lighthouse/lighthouse_geometry.c#L169-L201). 

While those are helpful places to look, my main issue has been struggles with the debugger and how it relates to the Kalman task. I'm still unable to get through basic debugging on the Kalman filter (i.e. viewing memory and how different variables change with various operations). For whatever reason, the Kalman filter seems to only be called with Gyroscope and Accelerometer measurement types instead of Sweep Angles [measurement_t](https://github.com/bitcraze/crazyflie-firmware/blob/master/src/modules/src/estimator/estimator_kalman.c#L298).
![Screenshot from 2023-08-07 17-47-09](https://github.com/bitcraze/discussions/assets/49169367/2abd9f6e-9871-46e4-a370-3f54855e4be5)

Another issue I've run into is viewing memory. By manually changing the measurement type with the debugger, I was finally able to get into the `kalmanCoreUpdateWithSweepAngles` [here](kalmanCoreUpdateWithSweepAngles). I was able to see important variables like rotation matrices within the `this.Rcf_` variable, but they kept getting reinitialized and turned out to be less useful.
![Screenshot from 2023-08-07 17-46-40](https://github.com/bitcraze/discussions/assets/49169367/c596f714-605f-4a57-a0e4-60b65ee6373b)

I guess what I'm getting at is that I'm new at this stuff, and I'm hoping for some help. I'm a student with little experience who has spent a huge portion of time on trying to understand this debugging with  little luck. Any help on debugging advice, would be greatly appreciated!!!

**Comment by knmcguire** [comment]
Hi!

I'm not sure if using a debugger to debug the kalman filter is the right approach here. The measurement for the gyro and accelerometers are polled at almost the same speed as the STM is running namely a 1000 hz. The basestation sweeps are only received at 50 hz times the amount of basestations, so it makes sense that you will only see the IMU measurements if you only try a few stops every time. It feels very inefficient...

Before going into this, why would you want gdb debugging exactly? What are the values you are most interested in and what would you like to achieve? I'm assuming you'd like to provide more orientation than just the 'on top flat face' lighthouse deck that is currently a bit hardcoded in the crazyflie lighthouse sweep code.

**Reply by willgraham345** [reply]
Ah, that makes sense. I was wondering why I couldn't ever get it to go into the loop. It confused me even more when there was values within it, and I couldn't seem to get back in. 

Yes you are right, I would like to provide another orientation for the lighthouse deck in mounting the lighthouse deck from below. Our configuration will require the lighthouses to be situation below the crazyflies. That's the only reason I'm digging into the Kalman filter and attempting to change/debug different rotation matrices and position vectors (like `Rcf_` and `s_` and `sr`). Beyond adding an additional orientation, I don't have any need to modify the Kalman Filter.

My other goal is to provide additional functionality so that the Lighthouse configuration could be written by an outside ROS2 program. We would like to mount the Lighthouses on mobile robots, pointing up at the Crazyflies. Ideally, we'd like to have the Lighthouses have mutable configuration controlled by a service. We're working on the in's and out's of those algorithms now. If you have any information on pointing us in the right direction that would be very appreciated. 

Thank you for your help!!

**Reply by knmcguire** [reply]
That is a really cool application. If you succeed, I hope you'll be open for writing a blogpost at bitcraze.io :) But let's get something working first!

We are now with a skeleton crew at the office due to holidays so I'll need to find some time to look on how to solve this properly. Would it be okay for you if we are able to provide more detail next week?

In the mean time, I've opened up this old issue https://github.com/bitcraze/crazyflie-firmware/issues/1197. We have been considering this for a while but lacking usecases, but now we have heard of a few request already.

**Reply by willgraham345** [reply]
Of course!! I'd be more than happy to provide information, what I've tried already, and even arrange a zoom/chat meeting if that would be helpful! 


I'd be more than happy to write a blogpost! The goal of this is to eventually publish a paper, so writing a blogpost would be a simple and fun addition.

**Reply by knmcguire** [reply]
Sure! I would say that it might be better to setup a meeting after September when more of my colleagues are back, but if you can add information on that issue I've given before that would help us and others. 

Great! Just send us an email when you think you are ready. A blogpost would also be a good way to promote your paper's work as well :)

**Reply by willgraham345** [reply]
Perfect! I'll put together all that I've done within a Github repository and invite you as a collaborator so you can have access to my source code. I'll be sure to include a guide noting changes. 

Can't wait to start!

**Reply by willgraham345** [reply]
Wanted to reach out on this request, and say that I'd love to get a meeting on the schedule to coordinate efforts. Let me know what works best for Bitcraze and we can go from there!

**Reply by krichardsson** [reply]
I think this setting up the meeting is better to handle in an email, could you please drop us an email at contact@bitcraze.io.
Thanks!

**Reply by willgraham345** [reply]
Of course! I'll send that over now.

**Reply by willgraham345** [reply]
Email was sent, let me know if there were any issues on the receiving end. Thank you!!

**Reply by krichardsson** [reply]
It has been received, we're just a bit busy...

**Reply by willgraham345** [reply]
No worries at all! I had some issues earlier this week with an email not arriving. Just wanted to make sure that I would be ready when you are.
